compiler-plugin/src/main/scala/scalapb/compiler/ProtobufGenerator.scala

```diff
@@ -443,14 +452,32 @@ class ProtobufGenerator(val params: GeneratorParams) extends DescriptorPimps {
         fp.add(s"if (${toBaseType(field)(fieldNameSymbol)} != ${defaultValueForGet(field, true)}) { __size += ${sizeExpressionForSingleField(field, toBaseType(field)(fieldNameSymbol))} }")
       }
     } else if (field.isOptional) {
-      fp.add(s"if ($fieldNameSymbol.isDefined) { __size += ${sizeExpressionForSingleField(field, toBaseType(field)(fieldNameSymbol + ".get"))} }")
+      if (field.customSingleScalaTypeName.isDefined) {
+        fp.add(
+          s"if ($fieldNameSymbol.isDefined) {",
+          s"  val __value = ${toBaseType(field)(fieldNameSymbol + ".get")}",
+          s"  __size += ${sizeExpressionForSingleField(field, "__value")}",
+          s"};"
+        )
+      } else {
+        fp.add(s"if ($fieldNameSymbol.isDefined) { __size += ${sizeExpressionForSingleField(field, toBaseType(field)(fieldNameSymbol + ".get"))} }")
+      }
     } else if (field.isRepeated) {
       val tagSize = CodedOutputStream.computeTagSize(field.getNumber)
       if (!field.isPacked) {
         Types.fixedSize(field.getType) match {
           case Some(size) => fp.add(s"__size += ${size + tagSize} * $fieldNameSymbol.size")
-          case None => fp.add(
-            s"$fieldNameSymbol.foreach($fieldNameSymbol => __size += ${sizeExpressionForSingleField(field, toBaseType(field)(fieldNameSymbol))})")
+          case None =>
+            if (field.customSingleScalaTypeName.isDefined) {
+              fp.add(
```

Comment:
Would it make sense to use this code for both cases (and then remove the `if`?) This would reduce duplication.

Reply:
I thought it through before I worked on it and decided to stay consistent with PR #333: basically trade-off code verbosity to avoid introducing any possible regression to the existing generated code for the more common cases where custom mappers are not used. 

Technically you're absolutely right there's no need for us to branch out for `field.customSingleScalaTypeName.isDefined` for any of these since the `if` block covers for both cases. I'm just not 100% sure whether the extra `val` assignment when it's only used once and there's no custom mappers involved would introduce any overhead. I'd assume JVM JIT is smart enough to make it irrelevant, just can't guarantee without actual measurement. Hence I chose the safer route.

Let me know if you still want to collapse both branches into just one to reduce code, I'll send in the commits.

Reply:
Let's collapse to a single code path. I'll be surprised if the JIT doesn't take care of this case.